spicy/toolchain/src/compiler/visitors/normalizer.cc

```diff
@@ -128,16 +128,48 @@ struct Visitor : public hilti::visitor::PostOrder<void, Visitor> {
         // Link hook to its unit type and field.
 
         NodeRef unit_type_ref = p.findParentRef<type::Unit>();
-        if ( ! unit_type_ref ) {
+        if ( unit_type_ref ) {
+            // Produce a tailored error message if `%XXX` is used on a unit field.
+            if ( auto id = h.id().namespace_(); id && hilti::util::startsWith(h.id().local(), "0x25_") ) {
+                if ( unit_type_ref->as<type::Unit>().itemByName(h.id().namespace_().local()) ) {
+                    p.node.addError(hilti::util::fmt("cannot use '%s' with a unit field",
+                                                     hilti::util::replace(h.id().local(), "0x25_", "%")));
+                    return;
+                }
+            }
+        }
+        else {
             // External hook, do name lookuo.
             auto ns = h.id().namespace_();
             if ( ! ns )
                 return;
 
             auto resolved = hilti::scope::lookupID<hilti::declaration::Type>(ns, p, "unit type");
             if ( ! resolved ) {
-                p.node.addError(resolved.error());
-                return;
+                // Look up as a type directly. If found, add explicit `%done`.
+                resolved = hilti::scope::lookupID<hilti::declaration::Type>(h.id(), p, "unit type");
+                if ( resolved ) {
+                    logChange(p.node, "adding explicit %done hook");
+                    p.node.as<Hook>().setID(h.id() + ID("0x25_done"));
+                    modified = true;
+                }
+                else {
+                    // Produce a tailored error message if `%XXX` is used on a unit field.
+                    if ( auto id = ns.namespace_(); id && hilti::util::startsWith(h.id().local(), "0x25_") ) {
+                        if ( auto resolved = hilti::scope::lookupID<hilti::declaration::Type>(id, p, "unit type") ) {
+                            if ( auto utype =
+                                     resolved->first->as<hilti::declaration::Type>().type().tryAs<type::Unit>();
+                                 utype && utype->itemByName(ns.local()) ) {
+                                p.node.addError(hilti::util::fmt("cannot use '%s' with a unit field",
+                                                                 hilti::util::replace(h.id().local(), "0x25_", "%")));
```

Comment:
Maybe more as a note, this will mutilate valid hook but unused hook names like `%done_0x25_foo` and show `%done_foo`. What we really want here is to replace a prefix `0x25_` with `%`, but we have no tooling for that. Same below.

Reply:
Yeah, not worth the trouble here I think, iirc we do this elsewhere too, and we control the names of hooks anyways.